allensdk/api/cloud_cache/cloud_cache.py

```diff
@@ -0,0 +1,525 @@
+        Reset output_stream to the beginning
+
+        Parameters
+        ----------
+        manifest_name: str
+            The name of the manifest to load. Must be an element in
+            self.manifest_file_names
+
+        output_stream: io.BytesIO
+            A byte stream into which to load the manifest
+
+    def _download_file(self, file_attributes: CacheFileAttributes) -> bool:
+
+        Check if a file exists and is in the expected state.
+
+        If it is, return True.
+
+        If it is not, download the file, creating the directory
+        where the file is to be stored if necessary.
+
+        If the download is successful, return True.
+
+        If the download fails (file hash does not match expectation),
+        return False.
+
+        Parameters
+        ----------
+        file_attributes: CacheFileAttributes
+            Describes the file to download
+
+        Returns
+        -------
+        None
+
+    """
+
+    _bucket_name = None
+
+    def __init__(self, cache_dir):
+        self._manifest = Manifest(cache_dir)
+        self._manifest_file_names = self._list_all_manifests()
+
+    @property
+    def file_id_column(self) -> str:
+        """
+        The column in the metadata files used to uniquely
+        identify data files
+        """
+        return self._manifest.file_id_column
+
+    @property
+    def version(self) -> str:
+        """
+        The version of the dataset currently loaded
+        """
+        return self._manifest.version
+
+    @property
+    def metadata_file_names(self) -> list:
+        """
+        List of metadata file names associated with this dataset
+        """
+        return self._manifest.metadata_file_names
+
+    @property
+    def manifest_file_names(self) -> list:
+        """
+        Sorted list of manifest file names associated with this
+        dataset
+        """
+        return copy.deepcopy(self._manifest_file_names)
+
+    def load_manifest(self, manifest_name: str):
+        """
+        Load a manifest from this dataset.
+
+        Parameters
+        ----------
+        manifest_name: str
+            The name of the manifest to load. Must be an element in
+            self.manifest_file_names
+        """
+        if manifest_name not in self.manifest_file_names:
+            raise ValueError(f"manifest: {manifest_name}\n"
+                             "is not one of the valid manifest names "
+                             "for this dataset:\n"
+                             f"{self.manifest_file_names}")
+
+        manifest_key = 'manifests/' + manifest_name
+        with io.BytesIO() as stream:
+            self._download_manifest(manifest_name, stream)
+            self._manifest.load(stream)
+
+    def _list_all_manifests(self) -> list:
+        """
+        Return a list of all of the file names of the manifests associated
+        with this dataset
+        """
+        raise NotImplementedError()
+
+    def _download_manifest(self,
```

Comment:
Using @abstractmethod decorator will force subclasses to implement these methods

Reply:
Is it considered reasonable practice to declare something an Abstract Base Class that isn't truly abstract (i.e. that does have some concrete method implementations in it)? Cursory googling of the question led me to believe the answer is "no".

Reply:
From my Java programming experience a long time ago, an abstract base class is one which cannot be instantiated and which has no concrete implementation. But a non-abstract class can have some abstract methods.

Reply:
I played around with the python. If the class doesn't inherit from `ABC` and the method is decorated with `abstractmethod`, you cannot call the method, but the "failure" is silent, i.e.
```
from abc import abstractmethod

class BaseClass(object):

    @abstractmethod
    def a(self):
        raise NotImplementedError()

    @abstractmethod
    def b(self):
        raise NotImplementedError()

    @abstractmethod
    def c(self):
        pass

class Daughter(BaseClass):
    def a(self):
        print('A')

    def b(self):
        print('B')

if __name__ == "__main__":

    x = Daughter()
    x.a()
    x.b()
    x.c()
```
results in 

```
(base) osxltFMD6W:abstract scott.daniel$ python test_abc.py 
A
B
```

I want the failure to make some noise, so I'll stick with Dan's suggestion that I just rely on the `NotImplementedError` (I'll add an informative error message)

Reply:
OK, did not know that the parent class had to inherit from `ABC` to force a subclass to implement the abstractmethod. 

If you don't want users to be able to instantiate this class, maybe it still makes sense to inherit from `ABC`.

Reply:
Also I just looked up that even Java allows abstract classes to contain both abstract as well as concrete implementations, so I think it is ok to inherit from `ABC` to not allow users to instantiate this class, if that's what you want.